### PR TITLE
Singleton map

### PR DESCRIPTION
**Pull request by cookfisher**

### CODE REVIEW

File: src/Cards/Cards.cpp [hunk 1]
```diff
@@ -162,12 +162,26 @@ Deck::~Deck() {
...
     auto card = deck.back();
-    deck.pop_back();
+    std::cout << "draw one card from the deck: " << *card << endl;
 
+    deck.pop_back();
+//    Check if the cards set shrinks after draw out one card and
+//    shuffles before each draw in order to simulate the random draw a card:
+    std::cout << "deck size: " << deck.size() << endl;
+//    std::cout << "next card: " << *deck.back() << endl;
+    std::cout << "slide the card to the end of Top Board" << endl;
```
Comment (mdesson) [inline src/Cards/Cards.cpp]: Please undo these changes.

File: src/Cards/Cards.cpp [hunk 2]
```diff
@@ -176,14 +190,18 @@ std::vector<Cards*> Deck::topBoardGenetor(Deck &deck) {
     for (auto i = 0; i < 6; i++) {
         tb->emplace_back(deck.draw());
     }
+//    std::cout << "TopBoard size: " << tb->size() << endl;
+    // displayTopBoard(*tb);
+//    std::cout << "topBoardGenetor(Deck deck) works" << endl;
```
Comment (mdesson) [inline src/Cards/Cards.cpp]: Why is this commented out?

File: src/Cards/Cards.cpp [hunk 3]
```diff
@@ -176,14 +190,18 @@ std::vector<Cards*> Deck::topBoardGenetor(Deck &deck) {
...
     return *tb;
 }
 
 void Deck::displayTopBoard(std::vector<Cards*> &topBoard) {
     int j = 0;
     int index =  0;
     for (auto it = topBoard.begin(); it != topBoard.end(); ++it)
-        cout << ++j << ") " << posArray[index++] << " Coins - " << **it << endl;
+        cout << "Position " << ++j << " - Cost: "<< posArray[index++] << " Coins. " << **it << endl;
+//    std::cout << "displayTopBoard( topBoard ) works" << endl;
```
Comment (mdesson) [inline src/Cards/Cards.cpp]: Revert these changes

File: src/Cards/Cards.cpp [hunk 4]
```diff
@@ -192,7 +210,14 @@ void Deck::updateTopBoard(int &position, std::vector<Cards*> &topBoard, Deck &de
...
     topBoard.emplace_back(deck.draw());
     cout << "Top Board cards series size is " << topBoard.size() << endl ;
+    std::cout << "updateTopBoard( topBoard ) works" << endl;
     displayTopBoard(topBoard);
+    std::cout << "----------------------"<<endl;
+//    auto tb = new std::vector<Cards*>();
+//    for (auto it = topBoard.begin(); it != topBoard.end(); ++it){
+//        tb->emplace_back(it);
+//    }
+//    return *tb;
```
Comment (mdesson) [inline src/Cards/Cards.cpp]: Revert these changes

File: src/Cards/Cards.cpp [hunk 5]
```diff
@@ -241,11 +266,21 @@ void Deck::exchange(Player &player, std::vector<Cards*> &topBoard, Deck &deck) {
...
     topBoard.emplace_back(deck.draw());
+    cout << "Top Board cards series size is " << topBoard.size() << endl ;
+    cout << player.getName() << " added the following card to their hand: " << *player.getGameHand()->back() << endl;
+//    cout << "Deck::exchange(hand, topboard, deck) works"  << endl;
+
+    // display updated top board
+    cout << "Top Board:" << endl;
+    displayTopBoard(topBoard);
     cout << player.getName() << " added the following card to their hand: " << *player.getGameHand()->back() << endl;
+    cout << endl << endl;
```
Comment (mdesson) [inline src/Cards/Cards.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 6]
```diff
@@ -4,229 +4,391 @@
 #include "GameState.h"
 #include "Map.h"
 
-void GameScore::generateWinner(std::vector<Player *> &players, Map &gameBoard) {
-    
-    auto regionOwners = gameBoard.getRegionOwners();
-    auto continentOwners = gameBoard.getContinentOwners();
+GameScore::GameScore() {}
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 7]
```diff
@@ -4,229 +4,391 @@
...
-        }
-        else if (players.size() == 5 && continentOwners.at(i).second == players.at(4)->getName()) {
-            players.at(4)->addPoints(1);
-        }
+GameScore::~GameScore() {
+    if (point) {
+        delete point;
+        point = NULL;
     }
+}
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 8]
```diff
@@ -4,229 +4,391 @@
...
 
-    // count points from hand
-    for (int i = 0; i < players.size(); i++) {
-        countPointsInHand(*players.at(i));
+int *GameScore::countTreePoint(int *tree) {
+    if (*tree < 2) {
+        point = new int(0);
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 9]
```diff
@@ -4,229 +4,391 @@
...
     }
-
-    // display final scores
-    std::cout << "Here are the final scores: " << endl;
-    for (int i = 0; i < players.size(); i++) {
-        std::cout << players.at(i)->getName() << ": " << players.at(i)->getPoints() << endl;
+    else if (*tree < 4) {
+        point = new int(1);
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 10]
```diff
@@ -4,229 +4,391 @@
...
-
-    // calculate winner
-    int winningPlayerIndex = 0;
-
-    for (int i = 0; i < players.size() - 1; i++) {
-        if (players.at(winningPlayerIndex)->getPoints() < players.at(i + 1)->getPoints()) {
-            winningPlayerIndex = i + 1;
-        }
+    else if (*tree < 5) {
+        point = new int(2);
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 11]
```diff
@@ -4,229 +4,391 @@
...
-            }
-            else if (wildExchange == "Carrot") {
-                carrots++;
-            }
-            else if (wildExchange == "Anvil") {
-                anvils++;
-            }
-        }
+    else if (*tree < 6) {
+        point = new int(3);
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 12]
```diff
@@ -4,229 +4,391 @@
...
-    crystals = countCrystalPoints(crystals);
-    anvils = countAnvilPoints(anvils);
-    carrots = countCarrotPoints(carrots);
-    trees = countTreePoints(trees);
-
-    player.addPoints(rocks + crystals + anvils + carrots + trees);
+    else {
+        point = new int(5);
+    }
+    return point;
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 13]
```diff
@@ -4,229 +4,391 @@
...
-        return 3;
+    else if (*anvil < 7) {
+        point = new int(3);
     }
     else {
-        return 5;
+        point = new int(5);
     }
+    return point;
 }
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 14]
```diff
@@ -4,229 +4,391 @@
...
-    else if (rock == 4) {
-        return 3;
+    else if (*crystal < 4) {
+        point = new int(3);
     }
     else {
-        return 5;
+        point = new int(5);
     }
+    return point;
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 15]
```diff
@@ -4,229 +4,391 @@
...
+            else if (*numCrystal > 0 && input == "crystal")
+                *numCrystal += 1;
+            else if (*numCarrot > 0 && input == "carrot"){
+                *numCarrot += 1;
+            }
+            else if (*numAnvil > 0 && input == "anvil")
+                *numAnvil += 1;
+            else
+                continue;
+        }
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 16]
```diff
@@ -4,229 +4,391 @@
...
+    return sum;
+}
+
+int *GameScore::numOfTree(std::vector<Cards*> &gameHand) {
+    int* num = new int(0); int* sum = new int(0);
+    for (int i = 0; i < gameHand.size(); i++) {
+         if (gameHand[i]->getGood()->compare("Tree") == 0) {
+            num = gameHand[i]->getNumOfGood();
+            *sum += *num;
+         }
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 17]
```diff
@@ -4,229 +4,391 @@
...
+    return sum;
+}
+
+int *GameScore::numOfAnvil(std::vector<Cards *> &gameHand) {
+    int* num = new int(0); int* sum = new int(0);
+    for (int i = 0; i < gameHand.size(); i++) {
+        if (gameHand[i]->getGood()->compare("Anvil") == 0) {
+            num = gameHand[i]->getNumOfGood();
+            *sum += *num;
+        }
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 18]
```diff
@@ -4,229 +4,391 @@
...
+    return sum;
+}
+
+int *GameScore::numOfCarrot(std::vector<Cards *> &gameHand) {
+    int* num = new int(0); int* sum = new int(0);
+    for (int i = 0; i < gameHand.size(); i++) {
+        if (gameHand[i]->getGood()->compare("Carrot") == 0) {
+            num = gameHand[i]->getNumOfGood();
+            *sum += *num;
+        }
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 19]
```diff
@@ -4,229 +4,391 @@
...
+    return sum;
+}
+
+int *GameScore::numOfRock(std::vector<Cards *> &gameHand) {
+    int* num = new int(0); int* sum = new int(0);
+    for (int i = 0; i < gameHand.size(); i++) {
+        if (gameHand[i]->getGood()->compare("Rock") == 0) {
+            num = gameHand[i]->getNumOfGood();
+            *sum += *num;
+        }
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 20]
```diff
@@ -4,229 +4,391 @@
...
+
+int *GameScore::numOfCrystal(std::vector<Cards *> &gameHand) {
+    int* num = new int(0); int* sum = new int(0);
+    for (int i = 0; i < gameHand.size(); i++) {
+        if (gameHand[i]->getGood()->compare("Crystal") == 0) {
+            num = gameHand[i]->getNumOfGood();
+            *sum += *num;
+        }
+    }
+    return sum;
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 21]
```diff
@@ -4,229 +4,391 @@
...
-int GameScore::countGoods(std::vector<Cards*> &gameHand, std::string good) {
-    int sum = 0;
+int *GameScore::numOfWild(std::vector<Cards *> &gameHand) {
+    int* num = new int(0); int* sum = new int(0);
     for (int i = 0; i < gameHand.size(); i++) {
-        if (*gameHand[i]->getGood() == good) {
-            sum += *gameHand[i]->getNumOfGood();
+        if (gameHand[i]->getGood()->compare("Wild") == 0) {
+            num = gameHand[i]->getNumOfGood();
+            *sum += *num;
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 22]
```diff
@@ -4,229 +4,391 @@
...
+        playerVector.at(i)->setArmiesOnBoard(p2);
+        cout << "  # armies on board : " << playerVector.at(i)->getArmiesOnBoard() << endl;
+        // count number of regions controlled
+        playerVector.at(i)->setRegions(p1);
+        cout <<"  # regions : "<< playerVector.at(i)->getRegions() << endl;
+        // count points from hand
+        int *p3 = computeGameScore(playerVector.at(i));
+        playerVector.at(i)->addPoints(*p3);
+        std::cout << "  hand points: " << *p3 << endl;
+        cout << "------------ regions and hand points added, # armies on board, # regions got ------------" << endl;
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.cpp [hunk 23]
```diff
@@ -4,229 +4,391 @@
...
+                    return;
+                }
+            }
         }
-    } while (invalidGood);
+    }
+}
 
-    return good;
-}
```
Comment (mdesson) [inline src/GameScore/GameScore.cpp]: Revert these changes

File: src/GameScore/GameScore.h [hunk 24]
```diff
@@ -7,21 +7,41 @@
...
+    int* numOfCrystal(std::vector<Cards*> &gameHand);
+
+    int* numOfWild(std::vector<Cards*> &gameHand);
+
+    int* computeGameScore(Player* &players);
+
+    void winnerGenerator ( std:: vector <Player*> &playerVector, Map &gameBoard);
 
-    // gameplay methods
-    void generateWinner (std::vector<Player*> &players, Map &gameBoard);
```
Comment (mdesson) [inline src/GameScore/GameScore.h]: Revert these changes

File: src/Map/Map.h [hunk 25]
```diff
@@ -39,12 +43,19 @@ class Map
...
     Map(std::vector<Node> inputNodes, std::string startRegion, std::vector<std::string> regions, std::vector<std::string> continents, std::vector<std::string> players);
     Map(Map const &m);
+
+public:
+
     ~Map();
 
+    // singleton instance
+    static Map* getInstance(std::vector<Node> inputNodes, std::string startRegion, std::vector<std::string> regions, std::vector<std::string> continents, std::vector<std::string> players);
+
```
Comment (mdesson) [inline src/Map/Map.h]: I'd overload `getInstance` so that we don't have to pass arguments to it every time we access the map

**Review by mdesson — CHANGES_REQUESTED** [review]
Can you add the folder `cmake-build-debug` to the `.gitignore`, no need to push it to github.

File: src/Map/Map.h [hunk 26]
```diff
@@ -39,12 +43,19 @@ class Map
     // recalculate the owner of a node
     void updateOwner(int index);
 
-public:
     // constructors and destructors
     Map(std::vector<Node> inputNodes, std::string startRegion, std::vector<std::string> regions, std::vector<std::string> continents, std::vector<std::string> players);
     Map(Map const &m);
```
Comment (mdesson) [inline src/Map/Map.h:48]: You can get rid of the copy constructor, it doesn't actually do anything anymore.

File: src/Map/Map.h [hunk 27]
```diff
@@ -39,12 +43,19 @@ class Map
...
+
+public:
+
     ~Map();
 
+    // singleton instance
+    static Map* getInstance(std::vector<Node> inputNodes, std::string startRegion, std::vector<std::string> regions, std::vector<std::string> continents, std::vector<std::string> players);
```
Comment (mdesson) [inline src/Map/Map.h:55]: I'd overload this as `static Map* getInstance()` so we don't need to give all these arguments each time we want to access the map.

File: src/Player/Player.cpp [hunk 28]
```diff
@@ -440,6 +440,10 @@ int Player::getAge() const { return *age; }
 
 int Player::getPoints() const { return *points; }
 
+int Player::getArmiesOnBoard() const { return armiesOnBoard; }
+
+int Player::getRegions() const { return regions; }
+
```
Comment (mdesson) [inline src/Player/Player.cpp]: Please revert, these methods are done in the Map.

File: src/Player/Player.cpp [hunk 29]
```diff
@@ -458,4 +462,11 @@ void Player::setAge(int age) { *Player::age = age; }
...
-void Player::addPoints(int additionalPoints) { *Player::points += additionalPoints; }
\ No newline at end of file
+void Player::addPoints(int additionalPoints) { *Player::points += additionalPoints; }
+
+void Player::setArmiesOnBoard(int &armiesOnBoard) { Player::armiesOnBoard = armiesOnBoard; }
+
+void Player::setRegions(int &regions) { Player::regions = regions; }
+
+
+
```
Comment (mdesson) [inline src/Player/Player.cpp]: Please revert, these methods are done in the Map.

File: src/Player/Player.h [hunk 30]
```diff
@@ -19,6 +19,8 @@ class Player
   int *cities;
   int *age;
   int *points;
+  int armiesOnBoard;
+  int regions;
```
Comment (mdesson) [inline src/Player/Player.h]: Please revert, these methods are done in the Map.

File: src/Player/Player.h [hunk 31]
```diff
@@ -51,6 +53,8 @@ class Player
   int getCities() const;
   int getAge() const;
   int getPoints() const;
+  int getArmiesOnBoard() const;
+  int getRegions() const;
```
Comment (mdesson) [inline src/Player/Player.h]: Please revert, these methods are done in the Map.

File: src/Player/Player.h [hunk 32]
```diff
@@ -62,6 +66,8 @@ class Player
   void setAge(int age);
   void setName(std::string name);
   void addPoints(int additionalPoints);
+  void setArmiesOnBoard(int &armiesOnBoard);
+  void setRegions(int &regions);
```
Comment (mdesson) [inline src/Player/Player.h]: Please revert, these methods are done in the Map.

**Review by mdesson — CHANGES_REQUESTED** [review]

**Review by seanheinrichs — COMMENTED** [review]
I have some reversions lined up to commit. The singleton stuff looks good, we can merge that as soon as the reversions are in this branch.
